ArduCopter/sensors.cpp

```diff
@@ -207,6 +207,62 @@ void Copter::read_battery(void)
     }
 }
 
+// Energy based battery failsafe calculations
+//  - Returns the legacy FS_BATT_MAH value if the FS_BATT_RESERVE parameter is 0 or RTL is not possible
+//  - Calculates watts to RTL based on estimated resting voltage and hover wattage values, adds reserve watts, returns calculated MAH
+float Copter::rtl_mah_calc(void)
+{
+    // If FS_BATT_RESERVE is disabled, or HOVER_WATT is 0, or no home position available, or EKF is bad, return the default MAH
+    if (!(g2.fs_batt_reserve) || !(g2.hover_watt) || ap.home_state == HOME_UNSET || !position_ok()) { 
+        return g.fs_batt_mah;
+    };
+
+    // Power required during various phases calculated as a percentage of hover power
+    uint16_t fs_rtl_watts_hover = g2.hover_watt;                                // hover watts
+    uint16_t fs_rtl_watts_climb = fs_rtl_watts_hover * g2.fs_batt_mult_clm;     // climb watts
+    uint16_t fs_rtl_watts_desc = fs_rtl_watts_hover * g2.fs_batt_mult_dsc;      // descent watts
+    uint16_t fs_rtl_watts_cruise = fs_rtl_watts_hover * g2.fs_batt_mult_crs;    // cruise watts
+
+    // Altitude calculations
+    int32_t target_alt_cm = MAX(current_loc.alt + g.rtl_climb_min, g.rtl_altitude); // Altitude copter will RTL at
+    int32_t climb_cm = target_alt_cm - current_loc.alt; // vertical distance to climb
+    int32_t desc_cm = target_alt_cm - LAND_START_ALT;   // vertical distance to descend at normal descent rate
+    int32_t final_desc_cm = LAND_START_ALT;             // vertical distance for final landing phase
+
+    // Watts and MAH vars
+    uint16_t fs_rtl_mah = g.fs_batt_mah;    // init mah just incase something goes wrong
+    uint16_t fs_rtl_watts = 0;              // init watts at zero
```

Comment:
It is a Watt*Hour unit, not Watts, right?

Reply:
Yes, watthours.